src/establish-metamask-repository.ts

```diff
@@ -0,0 +1,231 @@
+  const resolvedRepository = await resolveRepositoryReference({
+    repositoryReference,
+    workingDirectoryPath,
+    cachedRepositoriesDirectoryPath,
+  });
+
+  let defaultBranchName: string;
+  let lastFetchedDate: Date | null;
+  if (resolvedRepository.exists) {
+    log('Repository already exists at', resolvedRepository.directoryPath);
+    defaultBranchName = await getDefaultBranchName(
+      resolvedRepository.directoryPath,
+    );
+    lastFetchedDate = await getLastFetchedDate(
+      resolvedRepository.directoryPath,
+    );
+  } else {
+    ({ defaultBranchName, lastFetchedDate } = await cloneRepository({
+      repositoryShortname: resolvedRepository.shortname,
+      repositoryDirectoryPath: resolvedRepository.directoryPath,
+      cachedRepositoriesDirectoryPath,
+      outputLogger,
+    }));
+  }
+
+  return {
+    defaultBranchName,
+    lastFetchedDate,
+    shortname: resolvedRepository.shortname,
+    directoryPath: resolvedRepository.directoryPath,
+    createdAutomatically: resolvedRepository.createdAutomatically,
+  };
+}
+
+/**
+ * Clones a MetaMask repository using the `gh` tool.
+ *
+ * @param args - The arguments to this function.
+ * @param args.repositoryShortname - The name of the repository minus the
+ * organization (so, "utils" instead of "MetaMask/utils").
+ * @param args.repositoryDirectoryPath - The path where the repository should be
+ * cloned to.
+ * @param args.cachedRepositoriesDirectoryPath - The parent directory in which
+ * to keep the new repository.
+ * @param args.outputLogger - Writable streams for output messages.
+ * @returns Information about the cloned repository.
+ * @throws If `repositoryDirectoryPath` is not within
+ * `cachedRepositoriesDirectoryPath`, or you do not have `gh` installed.
+ */
+async function cloneRepository({
+  repositoryShortname,
+  repositoryDirectoryPath,
+  cachedRepositoriesDirectoryPath,
+  outputLogger,
+}: {
+  repositoryShortname: string;
+  repositoryDirectoryPath: string;
+  cachedRepositoriesDirectoryPath: string;
+  outputLogger: AbstractOutputLogger;
+}) {
+  /* istanbul ignore next: There's no way to reproduce this; this is just to be absolutely sure */
+  if (
+    path.dirname(repositoryDirectoryPath) !== cachedRepositoriesDirectoryPath
+  ) {
+    throw new Error(
+      'You seem to be pointing to a directory outside the cached repositories directory. Refusing to proceed to avoid data loss.',
+    );
+  }
+
+  log('Assuming', repositoryShortname, 'is the name of a MetaMask repo');
+
+  log('Removing existing', repositoryDirectoryPath);
+  await fs.promises.rm(repositoryDirectoryPath, {
+    recursive: true,
+    force: true,
+  });
+
+  log('Cloning', repositoryShortname, 'to', repositoryDirectoryPath);
+  outputLogger.logToStdout(
+    `Cloning repository MetaMask/${repositoryShortname}, please wait...`,
+  );
+  await ensureDirectoryStructureExists(cachedRepositoriesDirectoryPath);
+  // NOTE: This requires that you have `gh` installed locally
```

Comment:
Maybe worth considering doing something like this: https://github.com/MetaMask/action-tech-challenge-setup/blob/44f1f556a1faf04e3b7042c856f801a9fde2d952/scripts/setup.sh#L35

to ensure `gh` is present and authenticated.

Maybe not in this PR, but a future one? That would make more sense closer to entrypoint (maybe in the CLI layer), which doesn't exist yet here

Reply:
Yeah, absolutely!